README.md
# @CentralPing/http-redirect

[![Build Status](https://travis-ci.org/CentralPing/http-redirect.svg?branch=master)](https://travis-ci.org/CentralPing/http-redirect)
[![Coverage Status](https://coveralls.io/repos/github/CentralPing/http-redirect/badge.svg)](https://coveralls.io/github/CentralPing/http-redirect)
[![Dependency Status](https://david-dm.org/CentralPing/http-redirect.svg)](https://david-dm.org/CentralPing/http-redirect)
[![Greenkeeper Status](https://badges.greenkeeper.io/CentralPing/http-redirect.svg)](https://greenkeeper.io/)
[![Known Vulnerabilities](https://snyk.io/test/github/centralping/http-redirect/badge.svg)](https://snyk.io/test/github/centralping/http-redirect)

An HTTP request handler for node based [`http.Server`](https://nodejs.org/api/http.html#http_class_http_server) style servers. It automatically sets the correct status code and headers for a redirect response. Function invoctioners will still need to end the request manually.

## Notes

* The redirect URL is required.
* No safeguards against looped redirects.

## Installation

`npm i --save @centralping/http-redirect`

## API Reference

<a name="module_httpRedirect..redirect"></a>

### httpRedirect~redirect ⇒ <code>function</code>
Configures a request handler for `http.Server` style servers.

**Kind**: inner property of [<code>httpRedirect</code>](#module_httpRedirect)  
**Returns**: <code>function</code> - - `http.Server` style request handler  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| uri | <code>String</code> |  | the URI/URL to redirect HTTP request to. |
| [options] | <code>Object</code> |  | optional configuration for HTTP response. |
| [options.statusCode] | <code>Number</code> | <code>301</code> | the response HTTP status code. |
| [options.header] | <code>Object</code> |  | any additional valid HTTP response header values. |
| [options.header.cacheControl] | <code>String</code> | <code>&#x27;public, immutable, s-maxage&#x3D;31536000&#x27;</code> | the HTTP response header cache control. |


## Examples

### Simple

```js
const redirect = require('@centralping/http-redirect');

const redirector = redirect('https://foo.bar');

const server = someServerCreator();

// Redirect all the things
server.route('*', (req, res) => {
  // Do something interesting perhaps?
  // ...

  // Redirect this request
  redirector(req, res);

  // Important!!
  // End the response to send it on its merry way
  res.end();
});
```

### Middleware

```js
const redirect = require('@centralping/http-redirect');

const redirector = redirect('https://foo.bar');

const server = someServerCreator();

// Redirect all the things
server.route('*', redirector, (req, res) => res.end());
```

## Test

`npm test`

With coverage reporting:

`npm test -- --coverage`

With file watch:

`npm run watch`

## License

MIT


CHANGELOG.md
#### v0.0.2 (2018-11-02)

##### Chores

*  initial commit ([928bc489](https://github.com/CentralPing/http-redirect/commit/928bc489d16327b2fe616719d750dbde9cef341a))

##### Documentation Changes

* **readme:**  update description blurb ([b4cafc2a](https://github.com/CentralPing/http-redirect/commit/b4cafc2adeeda3ac7389f24239bf202844068132))

##### Tests

* **main:**  remove micro in favor of node http and remove chance of open handles ([46a24e65](https://github.com/CentralPing/http-redirect/commit/46a24e658f1352a05422461bf14315f30fd2beb9))



lib/main.js
// ESM syntax is supported.
// export {};

/**
 * @module httpRedirect
*/

/**
 * @name redirect
 * @description Configures a request handler for `http.Server` style servers.
 * @param {String} uri - the URI/URL to redirect HTTP request to.
 * @param {Object} [options] - optional configuration for HTTP response.
 * @param {Number} [options.statusCode=301] - the response HTTP status code.
 * @param {Object} [options.header] - any additional valid HTTP response header values.
 * @param {String} [options.header.cacheControl='public, immutable, s-maxage=31536000'] - the HTTP response header cache control.
 * @return {Function} - `http.Server` style request handler
 */
module.exports = (uri, {
  statusCode = 301,
  headers = {
    'Cache-Control': 'public, immutable, s-maxage=31536000'
  }
} = {}) => {
  if (uri === undefined) {
    throw new Error('A URL is required for redirects.');
  }

  const baseURI = uri.replace(/\/?$/, '');

  return ({url}, res) => {
    res.statusCode = statusCode;
    res.setHeader('Location', `${baseURI}${url}`);

    Object.entries(headers).forEach(([header, value]) => {
      res.setHeader(header, value);
    });
  };
};


lib/main.spec.unit.js
const {createServer} = require('http');
const request = require('request-promise-native');
const listen = require('test-listen');

const redirect = require('./main');

const testRequestOptions = {
  resolveWithFullResponse: true,
  followRedirect: false,
  // Don't reject messages that come back with error code (e.g. 404, 500s)
  simple: false
};

describe('[Unit] `redirect`', () => {
  beforeEach(() => jest.resetModules());

  it('should export a config function', () => {
    expect(redirect).toBeInstanceOf(Function);
  });

  it('should throw an error without a specified URL to redirect to', () => {
    expect(() => redirect()).toThrow('A URL is required for redirects.');
  });

  it('should return a configured function from config exection', () => {
    expect(redirect('http://foo')).toBeInstanceOf(Function);
  });

  it('should return redirect responses with defaults', async () => {
    const redirector = redirect('http://foo');
    const router = createServer((req, res) => {
      redirector(req, res);
      res.end();
    });
    const uri = await listen(router);
    const reqBase = request.defaults({baseUrl: uri});

    try {
      // Root request
      const resRoot = await reqBase({
        uri: '',
        ...testRequestOptions
      });
      expect(resRoot.statusCode).toBe(301);
      expect(resRoot.headers).toMatchObject({
        'location': 'http://foo/',
        'cache-control': 'public, immutable, s-maxage=31536000'
      });

      // Query parameters request
      const resQuery = await reqBase({
        uri: '',
        qs: {foo: 'bar'},
        ...testRequestOptions
      });
      expect(resQuery.statusCode).toBe(301);
      expect(resQuery.headers).toMatchObject({
        'location': 'http://foo/?foo=bar',
        'cache-control': 'public, immutable, s-maxage=31536000'
      });

      // Path request
      const resPath = await reqBase({
        uri: '/foo/bar',
        ...testRequestOptions
      });
      expect(resPath.statusCode).toBe(301);
      expect(resPath.headers).toMatchObject({
        'location': 'http://foo/foo/bar',
        'cache-control': 'public, immutable, s-maxage=31536000'
      });

      // Path and query parameter request
      const resPathQuery = await reqBase({
        uri: '/foo/bar',
        qs: {foo: 'bar'},
        ...testRequestOptions
      });
      expect(resPathQuery.statusCode).toBe(301);
      expect(resPathQuery.headers).toMatchObject({
        'location': 'http://foo/foo/bar?foo=bar',
        'cache-control': 'public, immutable, s-maxage=31536000'
      });

      await router.close();
    } catch (e) {
      await router.close();
      throw e;
    }
  });

  it('should return redirect responses with options', async () => {
    const redirector = redirect('http://foo', {
      statusCode: 302,
      headers: {'Cache-Control': 'no-cache'}
    });
    const router = createServer((req, res) => {
      redirector(req, res);
      res.end();
    });
    const uri = await listen(router);
    const reqBase = request.defaults({baseUrl: uri});

    try {
      // Root request
      const res = await reqBase({
        uri: '',
        ...testRequestOptions
      });
      expect(res.statusCode).toBe(302);
      expect(res.headers).toMatchObject({
        'location': 'http://foo/',
        'cache-control': 'no-cache'
      });

      await router.close();
    } catch (e) {
      await router.close();
      throw e;
    }
  });
});
